operator/v1alpha1/types_externaldns.go

```diff
@@ -0,0 +1,366 @@
+	Status ExternalDNSStatus `json:"status"`
+}
+
+// ExternalDNSSpec defines the desired state of the externalDNS.
+type ExternalDNSSpec struct {
+	// Domains specifies which domains that externalDNS should
+	// create DNS records for. Multiple domain union values
+	// can be specified such that sub domains of an included domain
+	// can effectively be ignored.
+	// Excluding any domains from domains that are not included
+	// will result in a no-op.
+	//
+	// +optional
+	Domains []ExternalDNSDomainUnion `json:"domains,omitempty"`
+
+	// Provider refers to the DNS provider that externalDNS
+	// should publish records to. Note that each externalDNS
+	// is tied to a single provider.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Provider ExternalDNSProvider `json:"provider"`
+
+	// Source describes which source resource
+	// externalDNS will be configured to create
+	// DNS records for.
+	//
+	// Multiple externalDNS resource would need to be
+	// created if multiple externalDNS source resources
+	// would be desired.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Source ExternalDNSSource `json:"source"`
+
+	// Zones describes which DNS Zone IDs
+	// externalDNS should publish records to.
+	//
+	// +optional
+	Zones []string `json:"zones,omitempty"`
+}
+
+// ExternalDNSDomainUnion describes how sets of included
+// or excluded domains are to be constructed.
+// +union
+type ExternalDNSDomainUnion struct {
+	// MatchType specifies the type of match to be performed
+	// by externalDNS when determining whether or not to publish DNS
+	// Records for a given source resource based on the resource's
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType,omitempty"`
+
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
+	// match a set of domains.
+	//
+	// +optional
+	Pattern *string `json:"pattern,omitempty"`
+
+	// FilterType marks the Name or Pattern field
+	// as an included or excluded set of domains.
+	//
+	// Note that excluding domains that
+	// have not been included will result
+	// in a no-op.
+	//
+	// This field accepts the following values:
+	//
+	//  "Include": Include the domain set specified
+	//  by name or pattern.
+	//
+	//  "Exclude": Exclude the domain set specified
+	//  by name or pattern.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	FilterType ExternalDNSFilterType `json:"filterType"`
```

Comment:
See https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-api-machinery/1027-api-unions/README.md, in particular the following: 

> All the fields must be optional (beside the discriminator) and will be included as members of the union.

and 

> - If there is a discriminator, and its value has changed, clear all fields but
>   the one specified by the discriminator,
> - If there is no discriminator, or if its value hasn't changed,
>   - if there is exactly one field, set the discriminator when there is one
>     to that value. Otherwise,
>   - compare the fields set before and after. If there is exactly one field
>     added, set the discriminator (if present) to that value, and remove all
>     other fields. if more than one field has been added, leave the process so
>     that validation will fail.

It isn't clear how the normalization is implemented, but it's in the spec, sooo...  We should avoid having non-union members (other than the discriminator) in the union type.  

Same applies to `ExternalDNSSource` below.  

We could get around this by defining a union type and inlining it alongside the non-union fields; see `IngressControllerCaptureHTTPCookie`, which inlines `IngressControllerCaptureHTTPCookieUnion`, for an example.  (We fixed the definition to use the inlining approach in https://github.com/openshift/api/pull/776/commits/d0822898eabb929c40c5146116252477abab8d18.)

Reply:
Thanks for the details! Fixed! :grin: